File: AppBoxTests/Common/MobileProvision/MobileProvisionTests.m
//
//  MobileProvisionTests.m
//  AppBoxTests
//
//  Created by AppBox on 29/05/26.
//  Copyright © 2026 Developer Insider. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "MobileProvision.h"

@interface MobileProvisionTests : XCTestCase
@property (nonatomic, strong) NSString *tempDir;
@end

@implementation MobileProvisionTests

- (void)setUp {
    [super setUp];
    self.tempDir = [NSTemporaryDirectory() stringByAppendingPathComponent:[[NSUUID UUID] UUIDString]];
    [[NSFileManager defaultManager] createDirectoryAtPath:self.tempDir withIntermediateDirectories:YES attributes:nil error:nil];
}

- (void)tearDown {
    [[NSFileManager defaultManager] removeItemAtPath:self.tempDir error:nil];
    [super tearDown];
}

- (NSString *)createProvisionFileWithPlist:(NSDictionary *)plist {
    NSData *plistData = [NSPropertyListSerialization dataWithPropertyList:plist format:NSPropertyListXMLFormat_v1_0 options:0 error:nil];
    NSString *plistString = [[NSString alloc] initWithData:plistData encoding:NSUTF8StringEncoding];
    // Wrap plist in binary content to simulate a real .mobileprovision file
    NSString *content = [NSString stringWithFormat:@"BINARY_HEADER_DATA%@BINARY_FOOTER_DATA", plistString];
    NSString *filePath = [self.tempDir stringByAppendingPathComponent:@"test.mobileprovision"];
    [content writeToFile:filePath atomically:YES encoding:NSISOLatin1StringEncoding error:nil];
    return filePath;
}

#pragma mark - Build Type Detection Tests

- (void)testBuildType_Enterprise_WhenProvisionsAllDevices {
    NSDictionary *plist = @{
        @"ProvisionsAllDevices": @YES,
        @"TeamIdentifier": @[@"ABC123"],
        @"TeamName": @"Enterprise Corp",
        @"UUID": @"test-uuid-enterprise"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertTrue(provision.isValid);
    XCTAssertEqualObjects(provision.buildType, BuildTypeEnterprise);
}

- (void)testBuildType_Development_WhenHasDevicesAndGetTaskAllow {
    NSDictionary *plist = @{
        @"ProvisionedDevices": @[@"UDID-1", @"UDID-2"],
        @"Entitlements": @{@"get-task-allow": @YES},
        @"TeamIdentifier": @[@"DEV123"],
        @"TeamName": @"Dev Team",
        @"UUID": @"test-uuid-dev"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertTrue(provision.isValid);
    XCTAssertEqualObjects(provision.buildType, BuildTypeDevelopment);
}

- (void)testBuildType_AdHoc_WhenHasDevicesButNoGetTaskAllow {
    NSDictionary *plist = @{
        @"ProvisionedDevices": @[@"UDID-1", @"UDID-2"],
        @"Entitlements": @{@"get-task-allow": @NO},
        @"TeamIdentifier": @[@"ADHOC123"],
        @"TeamName": @"AdHoc Team",
        @"UUID": @"test-uuid-adhoc"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertTrue(provision.isValid);
    XCTAssertEqualObjects(provision.buildType, BuildTypeAdHoc);
}

- (void)testBuildType_AppStore_WhenNoDevices {
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"STORE123"],
        @"TeamName": @"AppStore Team",
        @"UUID": @"test-uuid-appstore"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertTrue(provision.isValid);
    XCTAssertEqualObjects(provision.buildType, BuildTypeAppStore);
}

#pragma mark - Property Extraction Tests

- (void)testExtractsTeamId {
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"TEAM999"],
        @"TeamName": @"My Team",
        @"UUID": @"uuid-123"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.teamId, @"TEAM999");
}

- (void)testExtractsTeamName {
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"T1"],
        @"TeamName": @"Awesome Team",
        @"UUID": @"uuid-456"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.teamName, @"Awesome Team");
}

- (void)testExtractsUUID {
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"T1"],
        @"TeamName": @"Team",
        @"UUID": @"unique-profile-uuid"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.uuid, @"unique-profile-uuid");
}

- (void)testExtractsProvisionedDevices {
    NSArray *devices = @[@"AAA-BBB-CCC", @"DDD-EEE-FFF"];
    NSDictionary *plist = @{
        @"ProvisionedDevices": devices,
        @"Entitlements": @{@"get-task-allow": @YES},
        @"TeamIdentifier": @[@"T1"],
        @"TeamName": @"Team",
        @"UUID": @"uuid"
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.provisionedDevices, devices);
}

- (void)testExtractsCreationDate {
    NSDate *createDate = [NSDate dateWithTimeIntervalSince1970:1700000000];
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"T1"],
        @"TeamName": @"Team",
        @"UUID": @"uuid",
        @"CreationDate": createDate
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.createDate, createDate);
}

- (void)testExtractsExpirationDate {
    NSDate *expDate = [NSDate dateWithTimeIntervalSince1970:1800000000];
    NSDictionary *plist = @{
        @"TeamIdentifier": @[@"T1"],
        @"TeamName": @"Team",
        @"UUID": @"uuid",
        @"ExpirationDate": expDate
    };
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertEqualObjects(provision.expirationDate, expDate);
}

#pragma mark - Invalid Input Tests

- (void)testInvalidPath_IsNotValid {
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:@"/nonexistent/path.mobileprovision"];
    XCTAssertFalse(provision.isValid);
}

- (void)testFileWithNoPlist_IsNotValid {
    NSString *filePath = [self.tempDir stringByAppendingPathComponent:@"garbage.mobileprovision"];
    [@"this is not a valid mobileprovision file" writeToFile:filePath atomically:YES encoding:NSUTF8StringEncoding error:nil];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:filePath];
    XCTAssertFalse(provision.isValid);
}

- (void)testEmptyPlist_BuildType_DeveloperId {
    // An empty plist dict results in DeveloperID build type
    NSDictionary *plist = @{};
    NSString *path = [self createProvisionFileWithPlist:plist];
    MobileProvision *provision = [[MobileProvision alloc] initWithPath:path];
    XCTAssertTrue(provision.isValid);
    XCTAssertEqualObjects(provision.buildType, BuildTypeDeveloperId);
}

@end

File: AppBoxTests/Common/MobileProvision/MobileProvisionTests.swift
//
//  MobileProvisionTests.swift
//  AppBoxTests

import XCTest
@testable import AppBox

final class MobileProvisionTests: XCTestCase {

    private var tempDir = ""

    override func setUp() {
        super.setUp()
        tempDir = (NSTemporaryDirectory() as NSString).appendingPathComponent(UUID().uuidString)
        try? FileManager.default.createDirectory(atPath: tempDir, withIntermediateDirectories: true)
    }

    override func tearDown() {
        try? FileManager.default.removeItem(atPath: tempDir)
        super.tearDown()
    }

    private func makeProvisionFile(_ plist: [String: Any]) -> String {
        let data = try! PropertyListSerialization.data(fromPropertyList: plist, format: .xml, options: 0)
        let plistString = String(data: data, encoding: .utf8) ?? ""
        let content = "BINARY_HEADER_DATA\(plistString)BINARY_FOOTER_DATA"
        let filePath = (tempDir as NSString).appendingPathComponent("test.mobileprovision")
        try? content.write(toFile: filePath, atomically: true, encoding: .isoLatin1)
        return filePath
    }

    // MARK: - Build type detection

    func testBuildType_Enterprise_WhenProvisionsAllDevices() {
        let p = MobileProvision(path: makeProvisionFile([
            "ProvisionsAllDevices": true, "TeamIdentifier": ["ABC123"], "TeamName": "Enterprise Corp", "UUID": "test-uuid-enterprise"]))
        XCTAssertTrue(p.isValid)
        XCTAssertEqual(p.buildType, "enterprise")
    }

    func testBuildType_Development_WhenHasDevicesAndGetTaskAllow() {
        let p = MobileProvision(path: makeProvisionFile([
            "ProvisionedDevices": ["UDID-1", "UDID-2"], "Entitlements": ["get-task-allow": true],
            "TeamIdentifier": ["DEV123"], "TeamName": "Dev Team", "UUID": "test-uuid-dev"]))
        XCTAssertTrue(p.isValid)
        XCTAssertEqual(p.buildType, "development")
    }

    func testBuildType_AdHoc_WhenHasDevicesButNoGetTaskAllow() {
        let p = MobileProvision(path: makeProvisionFile([
            "ProvisionedDevices": ["UDID-1", "UDID-2"], "Entitlements": ["get-task-allow": false],
            "TeamIdentifier": ["ADHOC123"], "TeamName": "AdHoc Team", "UUID": "test-uuid-adhoc"]))
        XCTAssertTrue(p.isValid)
        XCTAssertEqual(p.buildType, "ad-hoc")
    }

    func testBuildType_AppStore_WhenNoDevices() {
        let p = MobileProvision(path: makeProvisionFile([
            "TeamIdentifier": ["STORE123"], "TeamName": "AppStore Team", "UUID": "test-uuid-appstore"]))
        XCTAssertTrue(p.isValid)
        XCTAssertEqual(p.buildType, "app-store")
    }

    // MARK: - Property extraction

    func testExtractsTeamId() {
        let p = MobileProvision(path: makeProvisionFile(["TeamIdentifier": ["TEAM999"], "TeamName": "My Team", "UUID": "uuid-123"]))
        XCTAssertEqual(p.teamId, "TEAM999")
    }

    func testExtractsTeamName() {
        let p = MobileProvision(path: makeProvisionFile(["TeamIdentifier": ["T1"], "TeamName": "Awesome Team", "UUID": "uuid-456"]))
        XCTAssertEqual(p.teamName, "Awesome Team")
    }

    func testExtractsUUID() {
        let p = MobileProvision(path: makeProvisionFile(["TeamIdentifier": ["T1"], "TeamName": "Team", "UUID": "unique-profile-uuid"]))
        XCTAssertEqual(p.uuid, "unique-profile-uuid")
    }

    func testExtractsProvisionedDevices() {
        let devices = ["AAA-BBB-CCC", "DDD-EEE-FFF"]
        let p = MobileProvision(path: makeProvisionFile([
            "ProvisionedDevices": devices, "Entitlements": ["get-task-allow": true],
            "TeamIdentifier": ["T1"], "TeamName": "Team", "UUID": "uuid"]))
        XCTAssertEqual(p.provisionedDevices, devices)
    }

    func testExtractsCreationDate() {
        let createDate = Date(timeIntervalSince1970: 1700000000)
        let p = MobileProvision(path: makeProvisionFile(["TeamIdentifier": ["T1"], "TeamName": "Team", "UUID": "uuid", "CreationDate": createDate]))
        XCTAssertEqual(p.createDate, createDate)
    }

    func testExtractsExpirationDate() {
        let expDate = Date(timeIntervalSince1970: 1800000000)
        let p = MobileProvision(path: makeProvisionFile(["TeamIdentifier": ["T1"], "TeamName": "Team", "UUID": "uuid", "ExpirationDate": expDate]))
        XCTAssertEqual(p.expirationDate, expDate)
    }

    // MARK: - Invalid input

    func testInvalidPath_IsNotValid() {
        XCTAssertFalse(MobileProvision(path: "/nonexistent/path.mobileprovision").isValid)
    }

    func testFileWithNoPlist_IsNotValid() {
        let filePath = (tempDir as NSString).appendingPathComponent("garbage.mobileprovision")
        try? "this is not a valid mobileprovision file".write(toFile: filePath, atomically: true, encoding: .utf8)
        XCTAssertFalse(MobileProvision(path: filePath).isValid)
    }

    func testEmptyPlist_BuildType_DeveloperId() {
        let p = MobileProvision(path: makeProvisionFile([:]))
        XCTAssertTrue(p.isValid)
        XCTAssertEqual(p.buildType, "developer-id")
    }
}
